tests/unit/s2n_hash_all_algs_test.c

```diff
@@ -0,0 +1,159 @@
+
+        RESULT_GUARD_POSIX(s2n_hash_new(&hash_copy));
+        RESULT_GUARD_POSIX(s2n_hash_copy(&hash_copy, hash_state));
+        RESULT_ENSURE_EQ(hash_copy.currently_in_hash, hash_state->currently_in_hash);
+        RESULT_ENSURE_EQ(hash_copy.is_ready_for_input, hash_state->is_ready_for_input);
+    }
+
+    /* Test s2n_hash_digest */
+    {
+        uint8_t digest_size = 0;
+        RESULT_GUARD_POSIX(s2n_hash_digest_size(hash_alg, &digest_size));
+        digest->size = digest_size;
+
+        RESULT_GUARD_POSIX(s2n_hash_digest(hash_state, digest->data, digest_size));
+        RESULT_ENSURE_EQ(hash_state->currently_in_hash, 0);
+        RESULT_ENSURE_EQ(hash_state->is_ready_for_input, false);
+
+        uint8_t copy_result[OUTPUT_DATA_SIZE] = { 0 };
+        RESULT_GUARD_POSIX(s2n_hash_digest(&hash_copy, copy_result, digest_size));
+        RESULT_ENSURE_EQ(hash_state->currently_in_hash, 0);
+        RESULT_ENSURE_EQ(hash_state->is_ready_for_input, false);
+        RESULT_ENSURE_EQ(memcmp(digest->data, copy_result, digest_size), 0);
+    }
+
+    RESULT_GUARD_POSIX(s2n_hash_free(&hash_copy));
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_hash_test(s2n_hash_algorithm hash_alg, struct s2n_blob *digest)
+{
+    struct s2n_hash_state hash_state = { 0 };
+
+    /* Test s2n_hash_new + s2n_hash_init */
+    {
+        RESULT_GUARD_POSIX(s2n_hash_new(&hash_state));
+        RESULT_ENSURE_EQ(hash_state.currently_in_hash, 0);
+        RESULT_ENSURE_EQ(hash_state.is_ready_for_input, false);
+
+        /* Allow MD5 when necessary */
+        if (s2n_is_in_fips_mode() && (hash_alg == S2N_HASH_MD5 || hash_alg == S2N_HASH_MD5_SHA1)) {
+            RESULT_GUARD_POSIX(s2n_hash_allow_md5_for_fips(&hash_state));
+        }
+
+        RESULT_GUARD_POSIX(s2n_hash_init(&hash_state, hash_alg));
+        RESULT_ENSURE_EQ(hash_state.currently_in_hash, 0);
+        RESULT_ENSURE_EQ(hash_state.is_ready_for_input, true);
+
+        RESULT_GUARD(s2n_hash_test_state(&hash_state, hash_alg, digest));
+    }
+
+    /* Test s2n_hash_reset */
+    {
+        struct s2n_blob result = { 0 };
+        uint8_t result_data[OUTPUT_DATA_SIZE] = { 0 };
+        RESULT_GUARD_POSIX(s2n_blob_init(&result, result_data, OUTPUT_DATA_SIZE));
+
+        RESULT_GUARD_POSIX(s2n_hash_reset(&hash_state));
+        RESULT_ENSURE_EQ(hash_state.currently_in_hash, 0);
+        RESULT_ENSURE_EQ(hash_state.is_ready_for_input, true);
+
+        RESULT_GUARD(s2n_hash_test_state(&hash_state, hash_alg, &result));
+        RESULT_ENSURE_EQ(digest->size, result.size);
+        RESULT_ENSURE_EQ(memcmp(digest->data, result.data, result.size), 0);
+    }
+
+    RESULT_GUARD_POSIX(s2n_hash_free(&hash_state));
+    return S2N_RESULT_OK;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    /* Calculate digests when not in FIPS mode. They must match. */
+    for (s2n_hash_algorithm hash_alg = 1; hash_alg < S2N_HASH_SENTINEL; hash_alg++) {
```

Comment:
I was trying to figure out why this is `1` and realized `S2N_HASH_NONE = 0`. It's redundant but do you think doing something like this could help?

```suggestion
    for (s2n_hash_algorithm hash_alg = S2N_HASH_NONE + 1; hash_alg < S2N_HASH_SENTINEL; hash_alg++) {
```

Reply:
I think you forgot to update the value here? after adding `S2N_HASH_NONE` to the test array, shouldn't it now be
```suggestion
    for (s2n_hash_algorithm hash_alg = S2N_HASH_NONE; hash_alg < S2N_HASH_SENTINEL; hash_alg++) {
```

Reply:
:facepalm: 

I've now updated it to use 0 directly, so that regardless of the value of S2N_HASH_NONE we're doing all the hash algs.